internal/service/cloud/network.go

```diff
@@ -170,7 +191,7 @@ func (s *Service) createLAN(ctx context.Context, ms *scope.Machine) error {
 		return fmt.Errorf("unable to create LAN in data center %s: %w", ms.DatacenterID(), err)
 	}
 
-	ms.ClusterScope.IonosCluster.SetCurrentRequestByDatacenter(ms.DatacenterID(),
+	ms.ClusterScope.SetCurrentRequestByDatacenter(ms.DatacenterID(),
 		http.MethodPost, sdk.RequestStatusQueued, requestPath)
 
 	err = ms.ClusterScope.PatchObject()
```

Comment:
I'm currently wondering if PatchObject should also be part of the SetCurrentRequestByDatacenter function.

Let's say we have 1 cluster with 2 machines.

Machine 1 in DC1
Machine 2 in DC2

Machine 1 creates LAN1 -> lock for DC1
Machine 2 creates LAN2 -> lock for DC2

Machine 1 sets current request for LAN1 in DC1 -> lock for Cluster
Machine 2 waits for unlock
Machine 1 unlocks the Cluster wide lock

Imagine the following two things happen at the same time:
* Machine 2 sets current request for LAN2 in DC2 -> lock for Cluster
* Machine 1 patches the current state of Cluster.Status

Machine 2 then patches again. 
We have a race condition here: Did Machine 1 patch both requests or only one request and what will happen to the patch request of machine 2 as both only know the initial cluster state, which was without any requests for a DC. Very unlikely but what if the patch of machine 2 finishes before machine 1?

In the end we expect to have two requests in the status then.

What do you think?

Reply:
> Very unlikely but what if the patch of machine 2 finishes before machine 1?

It's possible, so we have to think of it. And I did, but I didn't want to bloat the PR description too much. :D 

- First of all: Storing request IDs in the status is nice, but not essential. It's a shortcut to skip the request queue inspection. If we lose it, the controller will find the unfinished request nonetheless.
- `PatchObject` interally really only patches the calculated diff and at least for changes to `spec` and `status` it doesn't run into conflicts due to concurrent updates (that can only happen when updating conditions). The server will then merge the patched data with the whole resource.
- The described scenario is one of multiple possible race conditions.
- Concurrently adding entries is fine, they'll be merged, so your scenario is actually not a problem.
- However, entries can get lost when one actor attempts to delete the map's last entry while another adds to it. Deleting the last entry is done by setting the map to `null`. If that is processed after the request adding an entry, that entry is lost.
- The root of the problem is that the patch is calculated from the object as seen when building the patch helper in the beginning of the reconciliation. If the objects gets updated by someone else after that, the calculated diff is based on old data.
- Therefore, moving `PatchObject` into `SetCurrentRequestByDatacenter` will not help.
- I saw no obvious solution how to avoid potentially losing request URLs we'd like to have in the status as a shortcut. But as said initially, it's not a big deal.